src/core/src/core_logic/PatchInstaller.py

```diff
@@ -282,7 +333,142 @@ def install_updates(self, maintenance_window, package_manager, simulate=False):
+
+            packages_in_batch = []
+            package_versions_in_batch = []
+            skip_packages = []
+            already_installed_packages = []
+
+            for index in range(begin_index, end_index + 1):
+                if package_versions[index] == Constants.UA_ESM_REQUIRED:
+                    skip_packages.append(packages[index])
+                    self.status_handler.set_package_install_status(package_manager.get_product_name(packages[index]), str(package_versions[index]), Constants.NOT_SELECTED)
+                elif packages[index] not in self.last_still_needed_packages:
+                    # Should have got installed as dependent package of some other package. Package installation status should also have been set.
+                    already_installed_packages.append(packages[index])
+                    self.attempted_parent_package_install_count += 1
+                    self.successful_parent_package_install_count += 1
+                else:
+                    packages_in_batch.append(packages[index])
+                    package_versions_in_batch.append(package_versions[index])
+
+            if len(already_installed_packages) > 0:
+                self.composite_logger.log("Following packages are already installed. Should have got installed as dependent package of some other package " + str(already_installed_packages))
+
+            if len(skip_packages) > 0:
+                self.composite_logger.log("[Skipping packages " + str(skip_packages) + " - requires Ubuntu Advantage for Infrastructure with Extended Security Maintenance]")
+
+            if len(packages_in_batch) == 0:
+                continue
+
+            remaining_time = maintenance_window.get_remaining_time_in_minutes()
+
+            # point in time status
+            progress_status = self.progress_template.format(str(datetime.timedelta(minutes=remaining_time)), str(self.attempted_parent_package_install_count), str(self.successful_parent_package_install_count), str(self.failed_parent_package_install_count), str(installed_update_count - self.successful_parent_package_install_count),
+                                                            "Processing batch index: " + str(batch_index) + ", Number of packages: " + str(len(packages_in_batch)) + "\nProcessing packages: " + str(packages_in_batch))
+            self.composite_logger.log(progress_status)
+
+            if maintenance_window.is_package_install_time_available(remaining_time, len(packages_in_batch)) is False:
+                self.composite_logger.log("Stopped installing packages in batches as it is past the maintenance window cutoff time for installing in batches." +
+                                           " Batch Index: {0}, remaining time: {1}, number of packages in batch: {2}".format(batch_index, remaining_time, str(len(packages_in_batch))))
+                maintenance_window_batch_cutoff_reached = True
+                break
```

Comment:
Code flow in original code (sequential patching):
- calculate remaining time
- check is_package_install_time_available() and break if no time available 
- log progress_status 

Move #point in time status and the log below in to after maintenance window check

Maintenance window check should happen at the earliest possible level, and also the log statement in progress_status mentions we are processing packages, which starts only after maintenance window check has passed

Reply:
In the batch patching, following the order:

- calculate remaining time
- log progress_status
- check is_package_install_time_available() and break if no time available

log progress status is done before check is_package_install_time_available() so that actual value of metrics are logged before break due to no time available. 
If we do log progress status after maintenance window check then we do not get the final value of metrics and other details being logged in log progress status.
Please share your thoughts on this. Should we change the order in the existing code of installing packages sequentially to be in consistent with that in install packages in batch or change order in batch patching?

Reply:
Done the changes. Now following same order in both Sequential and Batch Patching:

- calculate remaining time
- check is_package_install_time_available() and break if no time available
- log progress_status